chain-impl-mockchain/src/leadership/genesis/mod.rs

```diff
@@ -295,6 +295,49 @@ mod tests {
         );
     }
 
+    #[test]
+    pub fn test_leader_with_invalid_pool_id() {
+        // Arrange
+        let leader_election_parameters = LeaderElectionParameters::new();
+
+        let cb = ConfigBuilder::new()
+            .with_slots_per_epoch(leader_election_parameters.slots_per_epoch)
+            .with_active_slots_coeff(leader_election_parameters.active_slots_coeff_as_milli());
+
+        let mut ledger = LedgerBuilder::from_config(cb)
+            .build()
+            .expect("cannot build test ledger")
+            .ledger;
+
+        let pools: HashMap<_, _>  = std::iter::from_fn(|| {
+            let (pool_id, pool_vrf_private_key) = make_pool(&mut ledger);
+            Some((
+                pool_id,
+                (pool_vrf_private_key, 0, leader_election_parameters.value),
+            ))
+        }).take(leader_election_parameters.pools_count).collect();
+
+        let selection = make_leadership_with_pools(&ledger, &pools);
+
+        let (invalid_pool_id, invalid_pool_vrf_private_key) = make_pool(&mut ledger);
+
+        // Act
+        let mut date = ledger.date();
+        let mut flag = true;
```

Comment:
I don't particularly like this approach with flag. maybe we could have a block like this:

```
let invalid_leader_elected_for_epoch = {
      for _ in 00..leader_election_parameters.slots_per_epoch{
            let invalid_leader = selection.leader(&invalid_pool_id, &invalid_pool_vrf_private_key, date).unwrap();
             if invalid_leader.is_some() {
                   return true;
            }
            date = date.next(ledger.era());
        }
        return false;
 }
 
 assert!(!invalid_leader_elected_for_epoch)
```

Reply:
In above it's clear what we are checking

Reply:
asserting on the return value of the `leader` function directly would be even better